### PR TITLE
Implementing Schedule Item PUD

### PR DESCRIPTION
**Pull request by alex-deaconu**
fix single item pud

### CODE REVIEW

File: src/main/java/com/risevision/viewer/client/controller/ViewerPlaceholderController.java [hunk 1]
```diff
@@ -434,14 +434,14 @@ else if (!gadgets.get(nextItem).isReady()) {
...
-			if (gadgets.size() != 1) {
+//			if (gadgets.size() != 1) {
 				if (!gadgets.get(currentItem).getItem().isPlayUntilDone()) {
 					int duration = RiseUtils.strToInt(gadgets.get(currentItem).getItem().getDuration(), 0);
 					// Schedule the timer to run once in x seconds.
 					setNextItemCheck(duration);
 //					itemTimer.schedule(duration * 1000);
 				}
-			}
+//			}
```
Comment (alex-deaconu) [inline src/main/java/com/risevision/viewer/client/controller/ViewerPlaceholderController.java]: Enables Item timer for playlists with just 1 item.

File: src/main/java/com/risevision/viewer/client/controller/ViewerPresentationController.java [hunk 2]
```diff
@@ -191,7 +191,7 @@ public void execute() {
 					// Send done command to the first placeholder
 					Command doneCommand = null;
 					if (placeholderDoneCommand != null && 
-							ViewerEntryPoint.isEmbed() &&
+//							ViewerEntryPoint.isEmbed() &&
 							(RiseUtils.strIsNullOrEmpty(presentation.getDonePlaceholder()) ||
 							presentation.getDonePlaceholder().equals(ph.getId()))) {
 						doneCommand = placeholderDoneCommand;
```
Comment (alex-deaconu) [inline src/main/java/com/risevision/viewer/client/controller/ViewerPresentationController.java]: Send done command to placeholder regardless of whether the Presentation is an Embed. Allows PUD to always be sent.

File: src/main/java/com/risevision/viewer/client/controller/ViewerScheduleController.java [hunk 3]
```diff
@@ -30,7 +30,7 @@
 	private ScheduleInfo schedule;
 	private ArrayList<PlaylistItemInfo> items = new ArrayList<PlaylistItemInfo>();
 	private ArrayList<ViewerPresentationController> presentations = new ArrayList<ViewerPresentationController>();
-	private int currentItem = -1, nextItem = -1, lastItem;
+	private int currentItem = -1, nextItem = -1, lastItem = -1;
```
Comment (alex-deaconu) [inline src/main/java/com/risevision/viewer/client/controller/ViewerScheduleController.java]: Fix issue where last item is initialized to `0` (first item) which is incorrect.

File: src/main/java/com/risevision/viewer/client/controller/ViewerScheduleController.java [hunk 4]
```diff
@@ -140,7 +140,7 @@ public void execute() {
 			scheduleReadyCommand.execute();
 		}
 		else if (presentations.size() == 1 && ViewerEntryPoint.isShowingProgressBar() && !ViewerEntryPoint.isEmbed()) {
-			timerController.startTimer();
+//			timerController.startTimer();
```
Comment (alex-deaconu) [inline src/main/java/com/risevision/viewer/client/controller/ViewerScheduleController.java]: Duplicate call, function is called in `play()`.

File: src/main/java/com/risevision/viewer/client/controller/ViewerScheduleController.java [hunk 5]
```diff
@@ -198,10 +198,27 @@ else if (presentations.get(i).getStatus() == ViewerPresentationController.UNLOAD
...
+
+//				nextItem = currentItem;
+				presentations.get(currentItem).play();
+				lastItem = currentItem;
+			}
+			else if (currentItem == -1) {
+				setNextItemCheck(-1);
+				
+				playNextItem();
+			}
```
Comment (alex-deaconu) [inline src/main/java/com/risevision/viewer/client/controller/ViewerScheduleController.java]: More comprehensive initialization scenario copied from ViewerPlaceholderController (https://github.com/Rise-Vision/viewer/blob/feature/schedule-item-pud/src/main/java/com/risevision/viewer/client/controller/ViewerPlaceholderController.java#L249)

File: src/main/java/com/risevision/viewer/client/controller/ViewerScheduleController.java [hunk 6]
```diff
@@ -221,11 +238,15 @@ private void onPresentationDone() {
...
-		else if (playing && currentItem != -1) {
+		else if (playing && currentItem != -1 && !items.get(currentItem).isPlayUntilDone() && 
+			lastItem == currentItem) {
 			ViewerEntryPoint.showBlackScreen(false);
 
 			presentations.get(currentItem).play();
 		}
+		else {
+			playNextItem();
+		}
```
Comment (alex-deaconu) [inline src/main/java/com/risevision/viewer/client/controller/ViewerScheduleController.java]: More comprehensive "done" logic copied from ViewerPlaceholderController (https://github.com/Rise-Vision/viewer/blob/feature/schedule-item-pud/src/main/java/com/risevision/viewer/client/controller/ViewerPlaceholderController.java#L265)

File: src/main/java/com/risevision/viewer/client/controller/ViewerScheduleController.java [hunk 7]
```diff
@@ -354,12 +375,29 @@ else if (ViewerEntryPoint.isDisplay() && !items.get(nextItem).getTimeline().canP
...
+				
+				setNextItemCheck(60);
+			}
+			else if (presentations.size() != 1) {
+				playNextItem();
+			}
+			else {
+				currentItem = nextItem;
+			}
+		}
```
Comment (alex-deaconu) [inline src/main/java/com/risevision/viewer/client/controller/ViewerScheduleController.java]: Added "isPlayUntilDone" initialization scenario which differs from non PUD items.

File: src/main/java/com/risevision/viewer/client/controller/ViewerScheduleController.java [hunk 8]
```diff
@@ -372,12 +410,14 @@ else if (presentations.get(nextItem).getStatus() != ViewerPresentationController
...
-//				itemTimer.schedule(duration * 1000);
-			}
+//			if (presentations.size() != 1) {
+				if (!items.get(currentItem).isPlayUntilDone()) {
+					int duration = RiseUtils.strToInt(items.get(currentItem).getDuration(), 0);
+					// Schedule the timer to run again in x seconds.
+					setNextItemCheck(duration);
+	//				itemTimer.schedule(duration * 1000);
+				}
+//			}
```
Comment (alex-deaconu) [inline src/main/java/com/risevision/viewer/client/controller/ViewerScheduleController.java]: Enable item timer for Schedules with only 1 Presentation. Consistent with above fix for Placeholders.

**Comment by alex-deaconu** [comment]
This is currently awaiting CS verification. Will deploy once that's done.

@tejohnso @ezequielc please review. Thanks.

**Comment by ezequielc** [comment]
@alex-deaconu PR looks good but I will defer to @tejohnso as I am not familiar with Viewer. Thanks

**Comment by tejohnso** [comment]
:+1:

**Pull request merged**